Add binding generator support for async iterable

youennf opened the pull request:
#### 10ce73287c2f5790bf83081440da55786d54ab6f
<pre>
Add binding generator support for async iterable
<a href="https://bugs.webkit.org/show_bug.cgi?id=231243">https://bugs.webkit.org/show_bug.cgi?id=231243</a>
rdar://problem/84158224

Reviewed by Yusuke Suzuki.

Refactor JSDOMAsyncIterator to better aligned with JSDOMIterator.
We in particular remove the need to know the JS wrapper type.
This allows to use more of the existing iterable binding generator support for async iterable.
Move FileSystemDirectoryHandle async iterable from custom code to binding generator code.

* LayoutTests/imported/w3c/web-platform-tests/file-system-access/idlharness.https.any-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/file-system-access/idlharness.https.any.worker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/file-system-access/sandboxed_FileSystemDirectoryHandle-getDirectoryHandle.https.any-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/file-system-access/sandboxed_FileSystemDirectoryHandle-getDirectoryHandle.https.any.worker-expected.txt:
* Source/WebCore/Modules/filesystemaccess/FileSystemDirectoryHandle.idl:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/bindings/js/ExtendedDOMIsoSubspaces.h:
* Source/WebCore/bindings/js/JSDOMAsyncIterator.h:
(WebCore::JSDOMAsyncIteratorBase::onPromiseSettledBody):
(WebCore::JSDOMAsyncIteratorBase::onPromiseFulFilledBody):
(WebCore::JSDOMAsyncIteratorBase::onPromiseRejectedBody):
(WebCore::IteratorTraits&gt;::destroy):
(WebCore::IteratorTraits&gt;::next):
(WebCore::IteratorTraits&gt;::runNextSteps):
(WebCore::IteratorTraits&gt;::getNextIterationResult):
(WebCore::IteratorTraits&gt;::settle):
(WebCore::IteratorTraits&gt;::createOnSettledFunction):
(WebCore::IteratorTraits&gt;::fulfill):
(WebCore::IteratorTraits&gt;::createOnFulfilledFunction):
(WebCore::IteratorTraits&gt;::reject):
(WebCore::IteratorTraits&gt;::createOnRejectedFunction):
(WebCore::IteratorTraits&gt;::finishCreation):
(WebCore::JSIterator&gt;::destroy): Deleted.
(WebCore::JSIterator&gt;::next): Deleted.
(WebCore::JSIterator&gt;::runNextSteps): Deleted.
(WebCore::JSIterator&gt;::getNextIterationResult): Deleted.
(WebCore::JSIterator&gt;::settle): Deleted.
(WebCore::JSIterator&gt;::fulfill): Deleted.
(WebCore::JSIterator&gt;::reject): Deleted.
(WebCore::JSIterator&gt;::finishCreation): Deleted.
* Source/WebCore/bindings/js/JSFileSystemDirectoryHandleCustom.cpp: Removed.
* Source/WebCore/bindings/js/JSFileSystemDirectoryHandleIterator.cpp: Removed.
* Source/WebCore/bindings/js/JSFileSystemDirectoryHandleIterator.h: Removed.
* Source/WebCore/bindings/scripts/CodeGeneratorJS.pm:
(GenerateInterface):
(AddAsyncIterableOperationIfNeeded):
(AddIterableOperationIfNeeded):
(addGenericIterableOperations):
(GetFunctionName):
(GeneratePropertiesHashTable):
(InterfaceNeedsAsyncIterator):
(GenerateImplementation):
(IsValueIterableInterface):
(IsKeyValueIterableInterface):
(GenerateAsyncIterableDefinition):
(GenerateIterableDefinition):
(GenerateGenericIterableDefinition):
(GetRuntimeEnabledStaticProperties):
* Source/WebCore/bindings/scripts/IDLParser.pm:
(parseAsyncIterable):
* Source/WebCore/bindings/scripts/preprocess-idls.pl:
(containsIterableInterfaceFromIDL):
* Source/WebCore/bindings/scripts/test/BindingTestGlobalConstructors.idl:
* Source/WebCore/bindings/scripts/test/JS/JSTestAsyncIterable.cpp: Added.
(WebCore::JSTestAsyncIterableDOMConstructor::prototypeForStructure):
(WebCore::JSTestAsyncIterableDOMConstructor::initializeProperties):
(WebCore::JSTestAsyncIterablePrototype::finishCreation):
(WebCore::JSTestAsyncIterable::JSTestAsyncIterable):
(WebCore::JSTestAsyncIterable::finishCreation):
(WebCore::JSTestAsyncIterable::createPrototype):
(WebCore::JSTestAsyncIterable::prototype):
(WebCore::JSTestAsyncIterable::getConstructor):
(WebCore::JSTestAsyncIterable::destroy):
(WebCore::JSC_DEFINE_CUSTOM_GETTER):
(WebCore::jsTestAsyncIterablePrototypeFunction_entriesCaller):
(WebCore::JSC_DEFINE_HOST_FUNCTION):
(WebCore::jsTestAsyncIterablePrototypeFunction_keysCaller):
(WebCore::jsTestAsyncIterablePrototypeFunction_valuesCaller):
(WebCore::TestAsyncIterableIteratorTraits&gt;::onPromiseSettled):
(WebCore::TestAsyncIterableIteratorTraits&gt;::onPromiseFulFilled):
(WebCore::TestAsyncIterableIteratorTraits&gt;::onPromiseRejected):
(WebCore::JSTestAsyncIterable::subspaceForImpl):
(WebCore::JSTestAsyncIterable::analyzeHeap):
(WebCore::JSTestAsyncIterableOwner::isReachableFromOpaqueRoots):
(WebCore::JSTestAsyncIterableOwner::finalize):
(WebCore::toJSNewlyCreated):
(WebCore::toJS):
(WebCore::JSTestAsyncIterable::toWrapped):
* Source/WebCore/bindings/scripts/test/JS/JSTestAsyncIterable.h: Added.
(WebCore::JSTestAsyncIterable::create):
(WebCore::JSTestAsyncIterable::createStructure):
(WebCore::JSTestAsyncIterable::subspaceFor):
(WebCore::wrapperOwner):
(WebCore::wrapperKey):
(WebCore::toJS):
(WebCore::toJSNewlyCreated):
* Source/WebCore/bindings/scripts/test/JS/JSTestAsyncKeyValueIterable.cpp: Added.
(WebCore::JSTestAsyncKeyValueIterableDOMConstructor::prototypeForStructure):
(WebCore::JSTestAsyncKeyValueIterableDOMConstructor::initializeProperties):
(WebCore::JSTestAsyncKeyValueIterablePrototype::finishCreation):
(WebCore::JSTestAsyncKeyValueIterable::JSTestAsyncKeyValueIterable):
(WebCore::JSTestAsyncKeyValueIterable::finishCreation):
(WebCore::JSTestAsyncKeyValueIterable::createPrototype):
(WebCore::JSTestAsyncKeyValueIterable::prototype):
(WebCore::JSTestAsyncKeyValueIterable::getConstructor):
(WebCore::JSTestAsyncKeyValueIterable::destroy):
(WebCore::JSC_DEFINE_CUSTOM_GETTER):
(WebCore::jsTestAsyncKeyValueIterablePrototypeFunction_entriesCaller):
(WebCore::JSC_DEFINE_HOST_FUNCTION):
(WebCore::jsTestAsyncKeyValueIterablePrototypeFunction_keysCaller):
(WebCore::jsTestAsyncKeyValueIterablePrototypeFunction_valuesCaller):
(WebCore::TestAsyncKeyValueIterableIteratorTraits&gt;::onPromiseSettled):
(WebCore::TestAsyncKeyValueIterableIteratorTraits&gt;::onPromiseFulFilled):
(WebCore::TestAsyncKeyValueIterableIteratorTraits&gt;::onPromiseRejected):
(WebCore::JSTestAsyncKeyValueIterable::subspaceForImpl):
(WebCore::JSTestAsyncKeyValueIterable::analyzeHeap):
(WebCore::JSTestAsyncKeyValueIterableOwner::isReachableFromOpaqueRoots):
(WebCore::JSTestAsyncKeyValueIterableOwner::finalize):
(WebCore::toJSNewlyCreated):
(WebCore::toJS):
(WebCore::JSTestAsyncKeyValueIterable::toWrapped):
* Source/WebCore/bindings/scripts/test/JS/JSTestAsyncKeyValueIterable.h: Added.
(WebCore::JSTestAsyncKeyValueIterable::create):
(WebCore::JSTestAsyncKeyValueIterable::createStructure):
(WebCore::JSTestAsyncKeyValueIterable::subspaceFor):
(WebCore::wrapperOwner):
(WebCore::wrapperKey):
(WebCore::toJS):
(WebCore::toJSNewlyCreated):
* Source/WebCore/bindings/scripts/test/JS/JSTestGlobalObject.cpp:
(WebCore::jsTestGlobalObject_TestAsyncIterableConstructorGetter):
(WebCore::JSC_DEFINE_CUSTOM_GETTER):
(WebCore::jsTestGlobalObject_TestAsyncKeyValueIterableConstructorGetter):
* Source/WebCore/bindings/scripts/test/JS/JSTestInterface.cpp:
(WebCore::jsTestInterfacePrototypeFunction_entriesCaller):
* Source/WebCore/bindings/scripts/test/JS/JSTestIterable.cpp:
(WebCore::jsTestIterablePrototypeFunction_entriesCaller):
* Source/WebCore/bindings/scripts/test/JS/JSTestNode.cpp:
(WebCore::jsTestNodePrototypeFunction_entriesCaller):
* Source/WebCore/bindings/scripts/test/SupplementalDependencies.dep:
* Source/WebCore/bindings/scripts/test/TestAsyncIterable.idl: Copied from Source/WebCore/bindings/js/ExtendedDOMIsoSubspaces.h.
* Source/WebCore/bindings/scripts/test/TestAsyncKeyValueIterable.idl: Copied from Source/WebCore/bindings/js/ExtendedDOMIsoSubspaces.h.

Canonical link: <a href="https://commits.webkit.org/257021@main">https://commits.webkit.org/257021@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/2f940a228c1d5bd9761ce350e3e057279b2f4ea9

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/97616 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/6865 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/30784 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/107112 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/167375 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/101567 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/7221 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/35626 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/90009 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/103771 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/103264 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/5403 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/84246 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/32411 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/87278 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/89076 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/75333 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/863 "Built successfully") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/20527 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/850 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/22005 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/4834 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/5670 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/60/builds/44469 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/2103 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/41397 "Passed tests") | | 
<!--EWS-Status-Bubble-End-->

Constellation commented on Source/WebCore/bindings/js/JSDOMAsyncIterator.h:
```diff
@@ -111,6 +108,21 @@ template<typename JSWrapper, typename IteratorTraits, typename JSIterator> class
+    JSC::EncodedJSValue settle(JSC::JSGlobalObject*);
+    JSC::JSBoundFunction* createOnSettledFunction(JSC::JSGlobalObject*);
+    static JSC::EncodedJSValue onPromiseSettledBody(JSC::JSGlobalObject* globalObject, JSC::CallFrame*, JSDOMAsyncIteratorBase* castedThis) { return castedThis->settle(globalObject); }
+    static JSC::EncodedJSValue onPromiseSettled(JSC::JSGlobalObject*, JSC::CallFrame*);
```
This does not work. All JS functions need to be defined via JSC_DECLARE_HOST_FUNCTION etc.. See JSDOMIterator::next's JSC_HOST_CALL_ATTRIBUTES and CodeGenerator.pm's JSC_ANNOTATE_HOST_FUNCTION for each next function.

Constellation commented on Source/WebCore/bindings/js/JSDOMAsyncIterator.h:
```diff
@@ -111,6 +108,21 @@ template<typename JSWrapper, typename IteratorTraits, typename JSIterator> class
+
+    JSC::EncodedJSValue fulfill(JSC::JSGlobalObject*, JSC::JSValue);
+    static JSC::EncodedJSValue onPromiseFulFilledBody(JSC::JSGlobalObject* globalObject, JSC::CallFrame* callFrame, JSDOMAsyncIteratorBase* castedThis) { return castedThis->fulfill(globalObject, callFrame->argument(0)); }
+    static JSC::EncodedJSValue onPromiseFulFilled(JSC::JSGlobalObject*, JSC::CallFrame*);
```
Ditto.

Constellation commented on Source/WebCore/bindings/js/JSDOMAsyncIterator.h:
```diff
@@ -111,6 +108,21 @@ template<typename JSWrapper, typename IteratorTraits, typename JSIterator> class
+
+    JSC::EncodedJSValue reject(JSC::JSGlobalObject*, JSC::JSValue);
+    static JSC::EncodedJSValue onPromiseRejectedBody(JSC::JSGlobalObject* globalObject, JSC::CallFrame* callFrame, JSDOMAsyncIteratorBase* castedThis) { return castedThis->reject(globalObject, callFrame->argument(0)); }
+    static JSC::EncodedJSValue onPromiseRejected(JSC::JSGlobalObject*, JSC::CallFrame*);
```
Ditto.

Constellation reviewed (changes requested):
Nice! Commented on bugs which causes crash on JIT Cage.

Constellation commented on Source/WebCore/bindings/js/JSDOMAsyncIterator.h:
```diff
@@ -291,21 +313,29 @@ JSC::EncodedJSValue JSDOMAsyncIteratorBase<JSWrapper, IteratorTraits, JSIterator
+}
+
+template<typename JSWrapper, typename IteratorTraits>
+JSC::EncodedJSValue JSC_HOST_CALL_ATTRIBUTES JSDOMAsyncIteratorPrototype<JSWrapper, IteratorTraits>::next(JSC::JSGlobalObject* globalObject, JSC::CallFrame* callFrame)
```
Also, if CodeGenerator.pm is generating this, then `JSC_ANNOTATE_HOST_FUNCTION` is necessary when materializing this next function. (As `JSFileSystemDirectoryHandleIteratorPrototypeNext` is doing right now).
See JSDOMIterator::next for example.

Constellation replied:
`JSC_ANNOTATE_HOST_FUNCTION` is still missing.

Constellation replied:
Or, maybe, need to update the output :)

Constellation replied:
Ah, wait, diff was not loaded. Looks like they are added now.

Constellation commented on Source/WebCore/bindings/scripts/test/JS/JSTestAsyncIterable.cpp:
```diff
@@ -0,0 +1,362 @@
+};
+
+using TestAsyncIterableIteratorPrototype = JSDOMAsyncIteratorPrototype<JSTestAsyncIterable, TestAsyncIterableIteratorTraits>;
+JSC_ANNOTATE_HOST_FUNCTION(TestAsyncIterableIteratorPrototypeNext, TestAsyncIterableIteratorPrototype::next);
```
This is right.

Constellation commented on Source/WebCore/bindings/scripts/test/JS/JSTestAsyncIterable.cpp:
```diff
@@ -0,0 +1,362 @@
+JSC::EncodedJSValue JSC_HOST_CALL_ATTRIBUTES TestAsyncIterableIteratorBase::onPromiseRejected(JSC::JSGlobalObject* globalObject, JSC::CallFrame* callFrame)
+{
+    return IDLOperation<TestAsyncIterableIteratorBase>::call<onPromiseRejectedBody>(*globalObject, *callFrame, "onPromiseRejected");
+}
```
They are not correct since they lack `JSC_ANNOTATE_HOST_FUNCTION`.

Constellation reviewed (approved):
r=me

youennf committed:
Add binding generator support for async iterable
https://bugs.webkit.org/show_bug.cgi?id=231243
rdar://problem/84158224

Reviewed by Yusuke Suzuki.

Refactor JSDOMAsyncIterator to better aligned with JSDOMIterator.
We in particular remove the need to know the JS wrapper type.
This allows to use more of the existing iterable binding generator support for async iterable.
Move FileSystemDirectoryHandle async iterable from custom code to binding generator code.

* LayoutTests/imported/w3c/web-platform-tests/file-system-access/idlharness.https.any-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/file-system-access/idlharness.https.any.worker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/file-system-access/sandboxed_FileSystemDirectoryHandle-getDirectoryHandle.https.any-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/file-system-access/sandboxed_FileSystemDirectoryHandle-getDirectoryHandle.https.any.worker-expected.txt:
* Source/WebCore/Modules/filesystemaccess/FileSystemDirectoryHandle.idl:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/bindings/js/ExtendedDOMIsoSubspaces.h:
* Source/WebCore/bindings/js/JSDOMAsyncIterator.h:
(WebCore::JSDOMAsyncIteratorBase::onPromiseSettledBody):
(WebCore::JSDOMAsyncIteratorBase::onPromiseFulFilledBody):
(WebCore::JSDOMAsyncIteratorBase::onPromiseRejectedBody):
(WebCore::IteratorTraits>::destroy):
(WebCore::IteratorTraits>::next):
(WebCore::IteratorTraits>::runNextSteps):
(WebCore::IteratorTraits>::getNextIterationResult):
(WebCore::IteratorTraits>::settle):
(WebCore::IteratorTraits>::createOnSettledFunction):
(WebCore::IteratorTraits>::fulfill):
(WebCore::IteratorTraits>::createOnFulfilledFunction):
(WebCore::IteratorTraits>::reject):
(WebCore::IteratorTraits>::createOnRejectedFunction):
(WebCore::IteratorTraits>::finishCreation):
(WebCore::JSIterator>::destroy): Deleted.
(WebCore::JSIterator>::next): Deleted.
(WebCore::JSIterator>::runNextSteps): Deleted.
(WebCore::JSIterator>::getNextIterationResult): Deleted.
(WebCore::JSIterator>::settle): Deleted.
(WebCore::JSIterator>::fulfill): Deleted.
(WebCore::JSIterator>::reject): Deleted.
(WebCore::JSIterator>::finishCreation): Deleted.
* Source/WebCore/bindings/js/JSFileSystemDirectoryHandleCustom.cpp: Removed.
* Source/WebCore/bindings/js/JSFileSystemDirectoryHandleIterator.cpp: Removed.
* Source/WebCore/bindings/js/JSFileSystemDirectoryHandleIterator.h: Removed.
* Source/WebCore/bindings/scripts/CodeGeneratorJS.pm:
(GenerateInterface):
(AddAsyncIterableOperationIfNeeded):
(AddIterableOperationIfNeeded):
(addGenericIterableOperations):
(GetFunctionName):
(GeneratePropertiesHashTable):
(InterfaceNeedsAsyncIterator):
(GenerateImplementation):
(IsValueIterableInterface):
(IsKeyValueIterableInterface):
(GenerateAsyncIterableDefinition):
(GenerateIterableDefinition):
(GenerateGenericIterableDefinition):
(GetRuntimeEnabledStaticProperties):
* Source/WebCore/bindings/scripts/IDLParser.pm:
(parseAsyncIterable):
* Source/WebCore/bindings/scripts/preprocess-idls.pl:
(containsIterableInterfaceFromIDL):
* Source/WebCore/bindings/scripts/test/BindingTestGlobalConstructors.idl:
* Source/WebCore/bindings/scripts/test/JS/JSTestAsyncIterable.cpp: Added.
(WebCore::JSTestAsyncIterableDOMConstructor::prototypeForStructure):
(WebCore::JSTestAsyncIterableDOMConstructor::initializeProperties):
(WebCore::JSTestAsyncIterablePrototype::finishCreation):
(WebCore::JSTestAsyncIterable::JSTestAsyncIterable):
(WebCore::JSTestAsyncIterable::finishCreation):
(WebCore::JSTestAsyncIterable::createPrototype):
(WebCore::JSTestAsyncIterable::prototype):
(WebCore::JSTestAsyncIterable::getConstructor):
(WebCore::JSTestAsyncIterable::destroy):
(WebCore::JSC_DEFINE_CUSTOM_GETTER):
(WebCore::jsTestAsyncIterablePrototypeFunction_entriesCaller):
(WebCore::JSC_DEFINE_HOST_FUNCTION):
(WebCore::jsTestAsyncIterablePrototypeFunction_keysCaller):
(WebCore::jsTestAsyncIterablePrototypeFunction_valuesCaller):
(WebCore::TestAsyncIterableIteratorTraits>::onPromiseSettled):
(WebCore::TestAsyncIterableIteratorTraits>::onPromiseFulFilled):
(WebCore::TestAsyncIterableIteratorTraits>::onPromiseRejected):
(WebCore::JSTestAsyncIterable::subspaceForImpl):
(WebCore::JSTestAsyncIterable::analyzeHeap):
(WebCore::JSTestAsyncIterableOwner::isReachableFromOpaqueRoots):
(WebCore::JSTestAsyncIterableOwner::finalize):
(WebCore::toJSNewlyCreated):
(WebCore::toJS):
(WebCore::JSTestAsyncIterable::toWrapped):
* Source/WebCore/bindings/scripts/test/JS/JSTestAsyncIterable.h: Added.
(WebCore::JSTestAsyncIterable::create):
(WebCore::JSTestAsyncIterable::createStructure):
(WebCore::JSTestAsyncIterable::subspaceFor):
(WebCore::wrapperOwner):
(WebCore::wrapperKey):
(WebCore::toJS):
(WebCore::toJSNewlyCreated):
* Source/WebCore/bindings/scripts/test/JS/JSTestAsyncKeyValueIterable.cpp: Added.
(WebCore::JSTestAsyncKeyValueIterableDOMConstructor::prototypeForStructure):
(WebCore::JSTestAsyncKeyValueIterableDOMConstructor::initializeProperties):
(WebCore::JSTestAsyncKeyValueIterablePrototype::finishCreation):
(WebCore::JSTestAsyncKeyValueIterable::JSTestAsyncKeyValueIterable):
(WebCore::JSTestAsyncKeyValueIterable::finishCreation):
(WebCore::JSTestAsyncKeyValueIterable::createPrototype):
(WebCore::JSTestAsyncKeyValueIterable::prototype):
(WebCore::JSTestAsyncKeyValueIterable::getConstructor):
(WebCore::JSTestAsyncKeyValueIterable::destroy):
(WebCore::JSC_DEFINE_CUSTOM_GETTER):
(WebCore::jsTestAsyncKeyValueIterablePrototypeFunction_entriesCaller):
(WebCore::JSC_DEFINE_HOST_FUNCTION):
(WebCore::jsTestAsyncKeyValueIterablePrototypeFunction_keysCaller):
(WebCore::jsTestAsyncKeyValueIterablePrototypeFunction_valuesCaller):
(WebCore::TestAsyncKeyValueIterableIteratorTraits>::onPromiseSettled):
(WebCore::TestAsyncKeyValueIterableIteratorTraits>::onPromiseFulFilled):
(WebCore::TestAsyncKeyValueIterableIteratorTraits>::onPromiseRejected):
(WebCore::JSTestAsyncKeyValueIterable::subspaceForImpl):
(WebCore::JSTestAsyncKeyValueIterable::analyzeHeap):
(WebCore::JSTestAsyncKeyValueIterableOwner::isReachableFromOpaqueRoots):
(WebCore::JSTestAsyncKeyValueIterableOwner::finalize):
(WebCore::toJSNewlyCreated):
(WebCore::toJS):
(WebCore::JSTestAsyncKeyValueIterable::toWrapped):
* Source/WebCore/bindings/scripts/test/JS/JSTestAsyncKeyValueIterable.h: Added.
(WebCore::JSTestAsyncKeyValueIterable::create):
(WebCore::JSTestAsyncKeyValueIterable::createStructure):
(WebCore::JSTestAsyncKeyValueIterable::subspaceFor):
(WebCore::wrapperOwner):
(WebCore::wrapperKey):
(WebCore::toJS):
(WebCore::toJSNewlyCreated):
* Source/WebCore/bindings/scripts/test/JS/JSTestGlobalObject.cpp:
(WebCore::jsTestGlobalObject_TestAsyncIterableConstructorGetter):
(WebCore::JSC_DEFINE_CUSTOM_GETTER):
(WebCore::jsTestGlobalObject_TestAsyncKeyValueIterableConstructorGetter):
* Source/WebCore/bindings/scripts/test/JS/JSTestInterface.cpp:
(WebCore::jsTestInterfacePrototypeFunction_entriesCaller):
* Source/WebCore/bindings/scripts/test/JS/JSTestIterable.cpp:
(WebCore::jsTestIterablePrototypeFunction_entriesCaller):
* Source/WebCore/bindings/scripts/test/JS/JSTestNode.cpp:
(WebCore::jsTestNodePrototypeFunction_entriesCaller):
* Source/WebCore/bindings/scripts/test/SupplementalDependencies.dep:
* Source/WebCore/bindings/scripts/test/TestAsyncIterable.idl: Copied from Source/WebCore/bindings/js/ExtendedDOMIsoSubspaces.h.
* Source/WebCore/bindings/scripts/test/TestAsyncKeyValueIterable.idl: Copied from Source/WebCore/bindings/js/ExtendedDOMIsoSubspaces.h.

Canonical link: https://commits.webkit.org/257021@main
